File: Demo/LGSideMenuControllerDemo/TableViewController.m
//
//  TableViewController.m
//  LGSideMenuControllerDemo
//
//  Created by Grigory Lutkov on 05.11.15.
//  Copyright (c) 2015 Grigory Lutkov. All rights reserved.
//

#import "TableViewController.h"
#import "MainViewController.h"
#import "NavigationController.h"
#import "ViewController.h"

@interface TableViewController ()

@property (strong, nonatomic) NSArray *titlesArray;

@end

@implementation TableViewController

- (id)init
{
    self = [super initWithStyle:UITableViewStylePlain];
    if (self)
    {
        self.title = @"LGSideMenuController";

        _titlesArray = @[@"Style Scale From Big",
                         @"Style Slide Above",
                         @"Style Slide Below",
                         @"Style Scale From Little",
                         @"Landscape always visible",
                         @"Status bar always visible",
                         @"Status bar light content",
                         @"Custom style"];

        [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"cell"];

        self.clearsSelectionOnViewWillAppear = NO;
    }
    return self;
}

#pragma mark - UITableView DataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return _titlesArray.count;
}

#pragma mark - UITableView Delegate

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];

    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;

    cell.textLabel.font = [UIFont systemFontOfSize:16.f];
    cell.textLabel.text = _titlesArray[indexPath.row];

    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 44.f;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    ViewController *viewController = [ViewController new];
    NavigationController *navigationController = [[NavigationController alloc] initWithRootViewController:viewController];
    MainViewController *mainViewController = nil;

    if (indexPath.row == 0)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:LGSideMenuPresentationStyleScaleFromBig
                                                                               type:0];
    else if (indexPath.row == 1)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:LGSideMenuPresentationStyleSlideAbove
                                                                               type:0];
    else if (indexPath.row == 2)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:LGSideMenuPresentationStyleSlideBelow
                                                                               type:0];
    else if (indexPath.row == 3)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:LGSideMenuPresentationStyleScaleFromLittle
                                                                               type:0];
    else if (indexPath.row == 4)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:LGSideMenuPresentationStyleScaleFromBig
                                                                               type:1];
    else if (indexPath.row == 5)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:LGSideMenuPresentationStyleSlideAbove
                                                                               type:2];
    else if (indexPath.row == 6)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:LGSideMenuPresentationStyleSlideAbove
                                                                               type:3];
    else if (indexPath.row == 7)
        mainViewController = [[MainViewController alloc] initWithRootViewController:navigationController
                                                                  presentationStyle:0
                                                                               type:4];

    UIWindow *window = [UIApplication sharedApplication].delegate.window;

    window.rootViewController = mainViewController;

    [UIView transitionWithView:window
                      duration:0.3
                       options:UIViewAnimationOptionTransitionCrossDissolve
                    animations:nil
                    completion:nil];
}

@end

File: Demo/Swift/LGSideMenuControllerDemo/TableViewController.swift
//
//  TableViewController.swift
//  LGSideMenuControllerDemo
//

class TableViewController: UITableViewController {

    private let titlesArray = ["Style \"Scale From Big\"",
                               "Style \"Slide Above\"",
                               "Style \"Slide Below\"",
                               "Style \"Scale From Little\"",
                               "Blurred root view cover",
                               "Blurred side views covers",
                               "Blurred side views backgrounds",
                               "Landscape always visible",
                               "Status bar always visible",
                               "Gesture area full screen",
                               "Editable table view",
                               "Custom style"]

    init() {
        super.init(style: .plain)

        title = "LGSideMenuController"

        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
    }

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    // MARK: - UITableViewDataSource

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return titlesArray.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)

        cell.accessoryType = .disclosureIndicator
        cell.textLabel!.font = UIFont.systemFont(ofSize: 16.0)
        cell.textLabel!.text = titlesArray[indexPath.row]
        
        return cell
    }
    
    // MARK: - UITableViewDelegate

    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 44.0
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let viewController: UIViewController

        if (indexPath.row == self.titlesArray.count - 2) {
            viewController = OtherViewController()
        }
        else {
            viewController = ViewController()
        }

        let navigationController = NavigationController(rootViewController: viewController)
        
        let mainViewController = MainViewController()
        mainViewController.rootViewController = navigationController
        mainViewController.setup(type: UInt(indexPath.row))

        let window = UIApplication.shared.delegate!.window!!
        window.rootViewController = mainViewController

        UIView.transition(with: window, duration: 0.3, options: [.transitionCrossDissolve], animations: nil, completion: nil)
    }

}
